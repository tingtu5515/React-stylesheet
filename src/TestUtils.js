/**
 * @flow
 */

import {CSSClassRepr, CSSClassJoinRepr} from './CSSClassRepr';

export const snapshotSerializer = {
  // $ExpectError: ...
  print(val, serialize, _indent) {
    if (val instanceof CSSClassRepr) {
      const repr = serialize(val.repr);
      return repr.replace(/^Object {/, `${val.reprName} {`);
    } else if (val instanceof CSSClassJoinRepr) {
      const repr = serialize(val.styles);
      return repr.replace(/^Array \[/, 'StyleJoin [');
    } else {
      return serialize(val);
    }
  },

  // $ExpectError: ...
  test(val) {
    return val instanceof CSSClassRepr || val instanceof CSSClassJoinRepr;
  },
};
