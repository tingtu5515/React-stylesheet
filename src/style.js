/**
 * @flow
 */

import type {CSSPropertySet} from './CSSType';
import type {Variant} from './Stylesheet';

import React from 'react';

import Stylesheet from './Stylesheet';
import getComponentDisplayName from './getComponentDisplayName';

export type ComponentSpec = {
  displayName?: string;
  [name: string]: CSSPropertySet;
};

export default function style<T: string | ReactClass<*>>(
  ComponentDefault: T,
  spec: ComponentSpec
): T {

  let {
    displayName,
    ...stylesheetSpec
  } = spec;

  if (displayName == null) {
    displayName = getComponentDisplayName(ComponentDefault);
  }

  let stylesheet = new Stylesheet(displayName, stylesheetSpec);

  let StylesheetComponent = class extends React.Component {

    props: {
      variant?: Variant;
      Component?: T;
      className?: string;
    };

    render() {
      let {
        variant,
        className: extraClassName,
        Component = ComponentDefault,
        ...props
      } = this.props;

      let className = stylesheet.toClassName(variant);
      if (extraClassName) {
        className = className + ' ' + extraClassName;
      }
      return (
        <Component
          {...props}
          className={className}
          />
      );
    }

    componentWillMount() {
      stylesheet.inject();
    }

    componentWillUnmount() {
      stylesheet.dispose();
    }
  };

  return ((StylesheetComponent: any): T);
}
