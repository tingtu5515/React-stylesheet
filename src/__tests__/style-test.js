import React from 'react';
import renderer from 'react-test-renderer';

import * as TestUtils from '../TestUtils';
import style from '../style';

expect.addSnapshotSerializer(TestUtils.snapshotSerializer);

test('render', function() {
  let SC = style('div', {
    base: {color: 'red'},
  });
  const tree = renderer.create(<SC>Hello</SC>).toJSON();
  expect(tree).toMatchSnapshot();
});

test('render w/ variant (inactive)', function() {
  let SC = style('div', {
    base: {color: 'red'},
    em: {fontWeight: 'bold'},
  });
  const tree = renderer.create(<SC>Hello</SC>).toJSON();
  expect(tree).toMatchSnapshot();
});

test('render w/ variant (active)', function() {
  let SC = style('div', {
    base: {color: 'red'},
    em: {fontWeight: 'bold'},
  });
  const tree = renderer.create(<SC variant={{em: true}}>Hello</SC>).toJSON();
  expect(tree).toMatchSnapshot();
});

test('styles host component with a stylesheet', function() {
  let SC = style('div', {base: {color: 'something'}});
  expect(SC.defaultProps.stylesheet).toBeTruthy();
  expect(SC.defaultProps.stylesheet.name).toBe('div');
  expect(SC.defaultProps.stylesheet.spec).toEqual({
    base: {
      color: 'something',
    },
  });
});

test('sets the displayName', function() {
  let SC = style('div', {
    displayName: 'custom',
    base: {
      color: 'something',
    },
  });
  expect(SC.defaultProps.stylesheet).toBeTruthy();
  expect(SC.defaultProps.stylesheet.name).toBe('custom');
  expect(SC.displayName).toBe('custom');
  expect(SC.defaultProps.stylesheet.spec).toEqual({
    base: {
      color: 'something',
    },
  });
});

test('styles composite component with a stylesheet', function() {
  function C(props) {
    return <div />;
  }
  let SC = style(C, {base: {color: 'something'}});
  expect(SC.defaultProps.stylesheet).toBeTruthy();
  expect(SC.defaultProps.stylesheet.name).toBe('C');
  expect(SC.defaultProps.stylesheet.spec).toEqual({
    base: {
      color: 'something',
    },
  });
});

test('overrides stylesheet of an already styled component', function() {
  let SC = style('div', {base: {color: 'something'}});
  expect(SC.defaultProps.stylesheet).toBeTruthy();
  expect(SC.defaultProps.stylesheet.spec).toEqual({
    base: {
      color: 'something',
    },
  });
  let SC2 = style('div', {
    displayName: 'custom',
    base: {
      color: 'something',
      hover: {
        color: 'red',
      },
    },
  });
  expect(SC2.displayName).toBe('custom');
  expect(SC2.defaultProps.stylesheet.name).toBe('custom');
  expect(SC2.defaultProps.stylesheet.spec).toEqual({
    base: {
      color: 'something',
      hover: {
        color: 'red',
      },
    },
  });
});
