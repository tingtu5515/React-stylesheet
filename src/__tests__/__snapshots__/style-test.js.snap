// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render 1`] = `
<div
  className={
    CompiledCSSWithVariants {
      "base": Object {
        "color": "red",
      },
    }
  }
>
  Hello
</div>
`;

exports[`render w/ variant (active) 1`] = `
<div
  className={
    CompiledCSSWithVariants {
      "base": Object {
        "color": "red",
      },
      "em": Object {
        "fontWeight": "bold",
      },
    }
  }
>
  Hello
</div>
`;

exports[`render w/ variant (inactive) 1`] = `
<div
  className={
    CompiledCSSWithVariants {
      "base": Object {
        "color": "red",
      },
    }
  }
>
  Hello
</div>
`;
